### PR TITLE
Add role banner under user avatar

### LINKED ISSUE #3393: Implement status icons in user avatars
As per the [spec on gamification](https://github.com/mozilla/pontoon/blob/3b840b967932e898b4ed398f7c7ea3d847106dd2/specs/0119-gamification-badges.md#status-icons), there should be four different status icons that users can have, which will be displayed next to their name:

- **New User Icon**: For users who joined less than 3 months ago.
- **Translator Icon**: For users contributing as translators.
- **Manager Icon**: For users serving as locale managers.
- **Admin Icon**: For users holding administrative roles.

For now, we've agreed to keep the icons as simple stylized text, and will iterate on the design as needed.

### PR DESCRIPTION
**Pull request by harmitgoswami**
Fix #3393 

This PR adds the user role as a banner under their avatar in the `translate` view. As per discussions in team calls, the design for the banner is to be iterated on within this PR.

**To reproduce results**: Head to any string within any team/project that has more than just the imported translation available (i.e. a user has submitted a translation for that string). At the bottom of their avatar, their role will be displayed within a banner.

image

### CODE REVIEW

File: pontoon/base/views.py [hunk 1]
```diff
@@ -414,12 +414,17 @@ def get_translation_history(request):
...
+            user_role = User.locale_role(t.user, locale)
+        else:
+            user_role = ""
         translation_dict.update(
             {
                 "user": u.name_or_email,
                 "uid": u.id,
                 "username": u.username,
                 "user_gravatar_url_small": u.gravatar_url(88),
+                "user_role": user_role,
```
Comment (mathjazz) [inline pontoon/base/views.py:427]: I believe the following should be the same (and then you can drop the new code above):
```suggestion
                "user_role": u.locale_role,
```

However, `User.locale_role()` is not exactly what we need here according to the spec. We also need the "New" icon and don't need the "System User" and "Contributor" roles.

File: translate/public/locale/en-US/translate.ftl [hunk 2]
```diff
@@ -704,6 +704,8 @@ user-UserAvatar--anon-alt-text =
     .alt = Anonymous User
 user-UserAvatar--alt-text = 
     .alt = User Profile
+user-UserAvatar--userRole =
+    { $userRole }
```
Comment (mathjazz) [inline translate/public/locale/en-US/translate.ftl:708]: This doesn't make the user roles localizable.

File: translate/src/modules/history/components/HistoryTranslation.tsx [hunk 3]
```diff
@@ -129,7 +129,7 @@ const User = ({
       {user}
     </a>
   ) : (
-    <span title={title}>{user}</span>
+    <span title={title}> {user} </span>
```
Comment (mathjazz) [inline translate/src/modules/history/components/HistoryTranslation.tsx:132]: Is there a reason for added whitespace?

**Review by mathjazz — COMMENTED** [review]
Nice work!

The spec is not exactly clear where to show the status icons. It mentions translations (suggestions) and comments and shows the status icon in the Profile. It also says "Status Icons: Show status icons universally in the user avatar".

I am not sure we should show them everywhere we show the avatar. In the Contributors dashboards and user filters, we already show the roles. I also don't think it's useful to show it in the Profile icon or in the Settings page, since that's only visible to the user.

How about we keep it just next to translations and comments (covered by this patch) and also add them to the Profile page? @flodolo Given that you were the reviewer of the spec, what's your take on this?

**Comment by flodolo** [comment]
> The spec is not exactly clear where to show the status icons. It mentions translations (suggestions) and comments and shows the status icon in the Profile. It also says "Status Icons: Show status icons universally in the user avatar".
> 
> I am not sure we should show them everywhere we show the avatar. In the Contributors dashboards and user filters, we already show the roles. I also don't think it's useful to show it in the Profile icon or in the Settings page, since that's only visible to the user.
> 
> How about we keep it just next to translations and comments (covered by this patch) and also add them to the Profile page? @flodolo Given that you were the reviewer of the spec, what's your take on this?

Looks like I didn't actually review the final version 🤔 

I agree on showing the status only in translations and comments:
* That information is useful when reviewing (is this person new? Is the person leaving a comment/suggestion a translator for that locale?)
* Same for comments (am I getting a comment from someone who doesn't know how Pontoon works?).

**Comment by harmitgoswami** [comment]
As we discussed, it seems like some of the statuses do indeed overflow when displayed in the banner. 

image

image

**Comment by flodolo** [comment]
Thoughts on using just letters to keep labels short (A/T/M + New), and rely on tooltip for explanation?

**Comment by harmitgoswami** [comment]
I think it would work for 'New', however, since a lot of users have their avatar as the default (the initials of their username), I personally feel that it might look odd to have up to three letters displayed in the avatar container.

**Comment by flodolo** [comment]
> I think it would work for 'New', however, since a lot of users have their avatar as the default (the initials of their username), I personally feel that it might look odd to have up to three letters displayed in the avatar container.

We could make it more "badge-y".

Screenshot 2024-10-08 alle 21 37 53

**Comment by mathjazz** [comment]
I like flod's idea for the status icon. 

Below is a rough variant of this proposal, which uses different colors for different statuses and shortens the titles using the similar method that we use on the dashboards.

Posnetek zaslona 2024-10-09 ob 10 09 03

File: pontoon/base/models/user.py [hunk 4]
```diff
@@ -192,6 +194,21 @@ def user_locale_role(self, locale):
         return "Contributor"
 
 
+def user_locale_status(self, locale):
+    if self is None:
```
[the remark this answers is not in the document]
Comment (mathjazz) [reply pontoon/base/models/user.py:198]: See https://github.com/mozilla/pontoon/pull/3400#discussion_r1792993102.

File: translate/src/modules/user/components/UserAvatar.tsx [hunk 5]
```diff
@@ -5,10 +5,13 @@ type Props = {
...
   title?: string;
   imageUrl: string;
+  userStatus?: string;
 };
 
 export function UserAvatar(props: Props): React.ReactElement<'div'> {
-  const { username, title, imageUrl } = props;
+  const { username, title, imageUrl, userStatus } = props;
+  const role = userStatus ? userStatus[0] : '';
+  const tooltip = userStatus ? userStatus[1] : '';
```
[the remark this answers is not in the document]
Comment (harmitgoswami) [reply translate/src/modules/user/components/UserAvatar.tsx:14]: Would `const [role, tooltip] = userStatus || [];` work instead? I had the suggested change before, but TypeScript seems to not like that :(

`Type 'string | undefined' must have a '[Symbol.iterator]()' method that returns an iterator.ts(2488)
`

[the remark this answers is not in the document]
Comment (mathjazz) [reply translate/src/modules/user/components/UserAvatar.tsx:14]: Can we instead make `userStatus` a required prop (by removing `?` in line 8)?

Right now the `AddComment` component doesn't use it, so we need to make it optional. But I'd show the user status there as well, given that it'll appear in the comment list after you post the comment.

Or at least it should. Seems like user statuses don't appear in the Comment list?

File: translate/src/modules/comments/components/Comment.tsx [hunk 6]
```diff
@@ -32,10 +32,13 @@ export function Comment(props: Props): null | React.ReactElement<'li'> {
 
   return (
     <li className='comment'>
-      <UserAvatar
-        username={comment.username}
-        imageUrl={comment.userGravatarUrlSmall}
-      />
+      <div className='avatar-container'>
```
[the remark this answers is not in the document]
Comment (harmitgoswami) [reply translate/src/modules/comments/components/Comment.tsx:36]: Sorry, I misread the original comment!

**Comment by harmitgoswami** [comment]
I've held off on creating `useLocaleRole.test.js` for now, in case we need to tweak the hook any further

File: translate/src/hooks/useLocaleRole.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+  if (!isAuthenticated) {
+    return ['', ''];
+  }
+
+  if (isAdmin) {
+    return ['ADMIN', 'Admin'];
+  }
+
+  if (Object.hasOwnProperty.call(managerForLocales, code)) {
```
Comment (mathjazz) [inline translate/src/hooks/useLocaleRole.ts:28]: I believe this is clearer:

```suggestion
  if (managerForLocales.includes(code)) {
```

File: translate/src/hooks/useLocaleRole.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    return ['MNGR', 'Manger'];
+  }
+
+  if (Object.hasOwnProperty.call(translatorForLocales, code)) {
```
Comment (mathjazz) [inline translate/src/hooks/useLocaleRole.ts:32]: Same as for `managerForLocales`.

File: translate/src/modules/comments/components/CommentsList.tsx [hunk 9]
```diff
@@ -18,6 +19,8 @@ export function CommentsList({
   user,
 }: Props): React.ReactElement<'div'> {
   const onAddComment = useAddCommentAndRefresh(translation);
+  const role = useLocaleRole();
```
Comment (mathjazz) [inline translate/src/modules/comments/components/CommentsList.tsx:22]: This seems unused, so changes to this file are not needed?

File: translate/src/modules/user/components/UserAvatar.tsx [hunk 10]
```diff
@@ -19,9 +21,19 @@ export function UserAvatar(props: Props): React.ReactElement<'div'> {
...
         onClick={(e: React.MouseEvent) => e.stopPropagation()}
       >
-        <Localized id='user-UserAvatar--alt-text' attrs={{ alt: true }}>
-          <img src={imageUrl} alt='User Profile' height='44' width='44' />
-        </Localized>
+        <div className='avatar-container'>
+          <Localized id='user-UserAvatar--alt-text' attrs={{ alt: true }}>
+            <img src={imageUrl} alt='User Profile' height='44' width='44' />
+          </Localized>
+          {userStatus && (
```
Comment (mathjazz) [inline translate/src/modules/user/components/UserAvatar.tsx:28]: Now that `userStatus` is always set, the check is no longer sufficient:

```suggestion
          {role && (
```

File: translate/src/modules/user/components/UserAvatar.tsx [hunk 11]
```diff
@@ -5,10 +5,12 @@ type Props = {
   username: string;
   title?: string;
   imageUrl: string;
+  userStatus: string | string[];
```
Comment (mathjazz) [inline translate/src/modules/user/components/UserAvatar.tsx:8]: `userStatus` should actually always be `string[]`:

```suggestion
  userStatus: string[];
```

To make the `make types` pass, you'll also need to fix the original sin - the type set in both files in the `api` folder.

File: translate/src/modules/user/components/UserAvatar.css [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+.user-avatar .avatar-container .user-status-banner.translator {
+  color: var(--status-translated);
+}
+
+.user-avatar .avatar-container .user-status-banner.new-user {
+  color: var(--status-fuzzy);
+}
+
+.history .translation:hover .user-status-banner {
+  border-color: var(--translation-border);
```
Comment (mathjazz) [inline translate/src/modules/user/components/UserAvatar.css:49]: Is this needed? Why only for translations?

File: translate/src/modules/user/components/UserAvatar.css [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+.history .translation .user-avatar img {
+  border-radius: 6px;
+  border: 2px solid var(--icon-border-1);
+}
+
+.history .translation:hover .user-avatar img {
+  border-color: var(--translation-border);
+}
```
Comment (mathjazz) [inline translate/src/modules/user/components/UserAvatar.css:59]: This two rules should be moved to the top, after the first rule.

**Review by mathjazz — COMMENTED** [review]
Works great! A few comments, but they are all small.

File: translate/src/modules/comments/components/CommentsList.tsx [hunk 14]
```diff
@@ -18,6 +18,7 @@ export function CommentsList({
   user,
 }: Props): React.ReactElement<'div'> {
   const onAddComment = useAddCommentAndRefresh(translation);
+
```
Comment (mathjazz) [inline translate/src/modules/comments/components/CommentsList.tsx:21]: Nit: Please restore this change and we can ship it! :)

**Review by mathjazz — APPROVED** [review]

**Pull request merged**